File: frontend/src/codeView/debugTypes.ts
export type DebugStatus =
  | "idle"
  | "starting"
  | "running"
  | "paused"
  | "stopping"
  | "terminated"
  | "error";

export type DebugConfigurationSummary = {
  name: string;
  type: string;
  request: string;
};

export type WorkspaceDebugSettings = {
  workspaceId: string;
  storagePath: string;
  revision: string;
  selectedConfiguration: string;
  configurations: DebugConfigurationSummary[];
  json: string;
  implicit: boolean;
};

export type DebugSourceBreakpoint = {
  line: number;
  column?: number;
};

export type DebugBreakpoint = {
  id?: number;
  path: string;
  line: number;
  column?: number;
  verified: boolean;
  message?: string;
};

export type DebugSourceLocation = {
  path?: string;
  name?: string;
  line?: number;
  column?: number;
  external?: boolean;
  sourceReference?: number;
};

export type DebugState = {
  workspaceId?: string;
  sessionId?: string;
  revision: number;
  status: DebugStatus;
  configuration?: string;
  adapterType?: string;
  threadId?: number;
  frameId?: number;
  currentLocation?: DebugSourceLocation;
  breakpoints?: DebugBreakpoint[];
  output?: string;
  error?: string;
  capabilities?: unknown;
};

export type DebugEvent = {
  workspaceId?: string;
  sessionId?: string;
  revision: number;
  type: string;
  state?: DebugState;
  category?: string;
  output?: string;
  message?: string;
};

export type DebugThread = { id: number; name: string };

export type DebugStackFrame = {
  id: number;
  name: string;
  location: DebugSourceLocation;
};

export type DebugScope = {
  name: string;
  presentationHint?: string;
  variablesReference: number;
  namedVariables?: number;
  indexedVariables?: number;
  expensive: boolean;
  location?: DebugSourceLocation;
};

export type DebugVariable = {
  name: string;
  value: string;
  type?: string;
  evaluateName?: string;
  variablesReference: number;
  namedVariables?: number;
  indexedVariables?: number;
  memoryReference?: string;
};

export type DebugThreadsResponse = {
  workspaceId: string;
  sessionId: string;
  revision: number;
  threads: DebugThread[];
};

export type DebugStackTraceResponse = {
  workspaceId: string;
  sessionId: string;
  revision: number;
  stackFrames: DebugStackFrame[];
  totalFrames?: number;
};

export type DebugScopesResponse = {
  workspaceId: string;
  sessionId: string;
  revision: number;
  scopes: DebugScope[];
};

export type DebugVariablesResponse = {
  workspaceId: string;
  sessionId: string;
  revision: number;
  variables: DebugVariable[];
};

export type DebugEvaluateResponse = {
  workspaceId: string;
  sessionId: string;
  revision: number;
  result: string;
  type?: string;
  variablesReference?: number;
  namedVariables?: number;
  indexedVariables?: number;
  memoryReference?: string;
};

File: internal/services/debug_types.go
package services

import "encoding/json"

const (
	DebugRuntimeEventName = debugEventName
	debugEventName        = "echo:debug:event"

	DebugStatusIdle       = "idle"
	DebugStatusStarting   = "starting"
	DebugStatusRunning    = "running"
	DebugStatusPaused     = "paused"
	DebugStatusStopping   = "stopping"
	DebugStatusTerminated = "terminated"
	DebugStatusError      = "error"
)

// DebugStartRequest selects and starts a workspace debug configuration. An
// empty configurationName uses the workspace selection (and, for a
// single-folder Go workspace, the implicit Go launch configuration).
type DebugStartRequest struct {
	ConfigurationName string `json:"configurationName,omitempty"`
	CurrentFile       string `json:"currentFile,omitempty"`
}

// DebugSessionRequest is used by execution-control operations.
type DebugSessionRequest struct {
	SessionID string `json:"sessionId"`
}

type DebugSourceBreakpoint struct {
	Line   int `json:"line"`
	Column int `json:"column,omitempty"`
}

type DebugBreakpoint struct {
	ID       int    `json:"id,omitempty"`
	Path     string `json:"path"`
	Line     int    `json:"line"`
	Column   int    `json:"column,omitempty"`
	Verified bool   `json:"verified"`
	Message  string `json:"message,omitempty"`
}

type DebugSetBreakpointsRequest struct {
	SessionID   string                  `json:"sessionId,omitempty"`
	SourcePath  string                  `json:"sourcePath"`
	Breakpoints []DebugSourceBreakpoint `json:"breakpoints"`
}

type DebugSourceLocation struct {
	Path      string `json:"path,omitempty"`
	Name      string `json:"name,omitempty"`
	Line      int    `json:"line,omitempty"`
	Column    int    `json:"column,omitempty"`
	External  bool   `json:"external,omitempty"`
	SourceRef int    `json:"sourceReference,omitempty"`
}

// DebugState is the recoverable public snapshot of the one app-wide debug
// session. References are valid only while this exact session is paused.
type DebugState struct {
	WorkspaceID     string               `json:"workspaceId,omitempty"`
	SessionID       string               `json:"sessionId,omitempty"`
	Revision        uint64               `json:"revision"`
	Status          string               `json:"status"`
	Configuration   string               `json:"configuration,omitempty"`
	AdapterType     string               `json:"adapterType,omitempty"`
	ThreadID        int                  `json:"threadId,omitempty"`
	FrameID         int                  `json:"frameId,omitempty"`
	CurrentLocation *DebugSourceLocation `json:"currentLocation,omitempty"`
	Breakpoints     []DebugBreakpoint    `json:"breakpoints,omitempty"`
	Output          string               `json:"output,omitempty"`
	Error           string               `json:"error,omitempty"`
	Capabilities    json.RawMessage      `json:"capabilities,omitempty"`
}

type DebugEvent struct {
	WorkspaceID string      `json:"workspaceId,omitempty"`
	SessionID   string      `json:"sessionId,omitempty"`
	Revision    uint64      `json:"revision"`
	Type        string      `json:"type"`
	State       *DebugState `json:"state,omitempty"`
	Category    string      `json:"category,omitempty"`
	Output      string      `json:"output,omitempty"`
	Message     string      `json:"message,omitempty"`
}

type DebugThreadsRequest struct {
	SessionID string `json:"sessionId"`
}

type DebugThread struct {
	ID   int    `json:"id"`
	Name string `json:"name"`
}

type DebugThreadsResponse struct {
	WorkspaceID string        `json:"workspaceId"`
	SessionID   string        `json:"sessionId"`
	Revision    uint64        `json:"revision"`
	Threads     []DebugThread `json:"threads"`
}

type DebugStackTraceRequest struct {
	SessionID  string `json:"sessionId"`
	ThreadID   int    `json:"threadId"`
	StartFrame int    `json:"startFrame,omitempty"`
	Levels     int    `json:"levels,omitempty"`
}

type DebugStackFrame struct {
	ID       int                 `json:"id"`
	Name     string              `json:"name"`
	Location DebugSourceLocation `json:"location"`
}

type DebugStackTraceResponse struct {
	WorkspaceID string            `json:"workspaceId"`
	SessionID   string            `json:"sessionId"`
	Revision    uint64            `json:"revision"`
	StackFrames []DebugStackFrame `json:"stackFrames"`
	TotalFrames int               `json:"totalFrames,omitempty"`
}

type DebugScopesRequest struct {
	SessionID string `json:"sessionId"`
	FrameID   int    `json:"frameId"`
}

type DebugScope struct {
	Name               string               `json:"name"`
	PresentationHint   string               `json:"presentationHint,omitempty"`
	VariablesReference int                  `json:"variablesReference"`
	NamedVariables     int                  `json:"namedVariables,omitempty"`
	IndexedVariables   int                  `json:"indexedVariables,omitempty"`
	Expensive          bool                 `json:"expensive"`
	Location           *DebugSourceLocation `json:"location,omitempty"`
}

type DebugScopesResponse struct {
	WorkspaceID string       `json:"workspaceId"`
	SessionID   string       `json:"sessionId"`
	Revision    uint64       `json:"revision"`
	Scopes      []DebugScope `json:"scopes"`
}

type DebugVariablesRequest struct {
	SessionID          string `json:"sessionId"`
	VariablesReference int    `json:"variablesReference"`
	Filter             string `json:"filter,omitempty"`
	Start              int    `json:"start,omitempty"`
	Count              int    `json:"count,omitempty"`
}

type DebugVariable struct {
	Name               string `json:"name"`
	Value              string `json:"value"`
	Type               string `json:"type,omitempty"`
	EvaluateName       string `json:"evaluateName,omitempty"`
	VariablesReference int    `json:"variablesReference"`
	NamedVariables     int    `json:"namedVariables,omitempty"`
	IndexedVariables   int    `json:"indexedVariables,omitempty"`
	MemoryReference    string `json:"memoryReference,omitempty"`
}

type DebugVariablesResponse struct {
	WorkspaceID string          `json:"workspaceId"`
	SessionID   string          `json:"sessionId"`
	Revision    uint64          `json:"revision"`
	Variables   []DebugVariable `json:"variables"`
}

type DebugEvaluateRequest struct {
	SessionID  string `json:"sessionId"`
	Expression string `json:"expression"`
	FrameID    int    `json:"frameId,omitempty"`
	Context    string `json:"context,omitempty"`
}

type DebugEvaluateResponse struct {
	WorkspaceID        string `json:"workspaceId"`
	SessionID          string `json:"sessionId"`
	Revision           uint64 `json:"revision"`
	Result             string `json:"result"`
	Type               string `json:"type,omitempty"`
	VariablesReference int    `json:"variablesReference,omitempty"`
	NamedVariables     int    `json:"namedVariables,omitempty"`
	IndexedVariables   int    `json:"indexedVariables,omitempty"`
	MemoryReference    string `json:"memoryReference,omitempty"`
}
